bionic/gcs.py

```diff
@@ -37,99 +35,33 @@ def get_gcs_client_without_warnings(cache_value=True):
+        # There seems to be a bug in fsspec due to which, the file is uploaded
+        # as the url, instead of inside the folder. What this means is, writing
+        # a file c.json to gs://a/b/ would result in file gs://a/b instead of
+        # gs://a/b/c.json.
+        #
+        # `put` API is supposed to write the file inside the folder but it strips
+        # the ending "/" at the end in fsspec's `_strip_protocol` method.
         if url.endswith("/"):
-            url = url[:-1]
-        self.url = url
-        bucket_name, object_prefix = bucket_and_object_names_from_gs_url(url)
-
-        self._bucket_name = bucket_name
-        self._object_prefix = object_prefix
-        self._init_client()
-
-    def __getstate__(self):
-        # Copy the object's state from self.__dict__ which contains
-        # all our instance attributes. Always use the dict.copy()
-        # method to avoid modifying the original state.
-        state = self.__dict__.copy()
-        # Remove the unpicklable entries.
-        del state["_client"]
-        del state["_bucket"]
-        return state
-
-    def __setstate__(self, state):
-        # Restore instance attributes.
-        self.__dict__.update(state)
-        # Restore the client and bucket.
-        self._init_client()
-
-    def _init_client(self):
-        self._client = get_gcs_client_without_warnings()
-        self._bucket = self._client.get_bucket(self._bucket_name)
-
-    def blob_from_url(self, url):
-        object_name = self._validated_object_name_from_url(url)
-        return self._bucket.blob(object_name)
-
-    def url_from_object_name(self, object_name):
-        return self._GS_URL_PREFIX + self._bucket.name + "/" + object_name
-
-    def blobs_matching_url_prefix(self, url_prefix):
-        obj_prefix = self._validated_object_name_from_url(url_prefix)
-        return self._bucket.list_blobs(prefix=obj_prefix)
-
-    @staticmethod
-    def gsutil_cp(src_url, dst_url):
-        gsutil_cp(src_url, dst_url)
-
-    def _validated_object_name_from_url(self, url):
-        bucket_name, object_name = bucket_and_object_names_from_gs_url(url)
-        assert bucket_name == self._bucket.name
-        assert object_name.startswith(self._object_prefix)
-        return object_name
-
-
-def gsutil_cp(src_url, dst_url):
-    """
-    This method is a proxy for _gsutil_cp which does the actual work. The proxy
-    exists so that _gsutil_cp can be replaced for testing.
-    """
-    _gsutil_cp(str(src_url), str(dst_url))
-
-
-def _gsutil_cp(src_url, dst_url):
-    args = [
-        "gsutil",
-        "-q",  # Don't log anything but errors.
-        "-m",  # Transfer files in parallel.
-        "cp",
-        "-r",  # Recursively sync sub-directories.
-        src_url,
-        dst_url,
-    ]
-    logger.debug("Running command: %s" % " ".join(args))
-    subprocess.check_call(args)
-    logger.debug("Finished running gsutil")
+            url = url + str(path).rsplit("/", 1)[1]
```

Comment:
As we discussed on Slack, `str(path).rsplit("/", 1)[1]` seems like it could just be `path.name`.